Make CircularBuffer conform to more protocols [Fixes #350]

esttorhe opened the pull request:
Make `CircularBuffer` conform to more protocols

### Motivation:

During the `PR` apple/swift-nio#347 there was a discussion around which protocols this auto-expanding buffer should conform; the following 3 were the winners:

1. `BidirectionalCollection`
1. `RandomAccessCollection`
1. `RangeReplaceableCollection`

### Modifications:

Added conformance to the 3 protocols and provided an implementation of 5 methods:

1. `replaceSubrange(:with:)`
1. `index(before:)`
1. `removeSubrange(:)`
1. `remove(at:)`
1. `removeLast(:)`


### Result:

With the conformance of this 3 protocols and the implementation of this 5 functions we get `dropLast()`, `dropLast(:)`, `removeFirst()`, `removeLast()` & `popLast()` for free.

### Comments:

I tried to make the updates of the internal `buffer` as direct as possible without going through lots of `API`s (e.g. directly manipulating the head and tail indexes whenever possible) to prevent lots of overhead.

Same with the `replaceSubrange` where I added a couple of checks that make the code more verbose (if the range and the count of new elements matches we only need to override values; or if the count of new elements is 0 that means we want to remove elements; etc).

I thought about adding a few extra scenarios (e.g: the `lowerBound` is `headIdx`; thus we only need to `nil` out the elements up to `upperBound` and update the reference to `headIdx`; and same case where `upperBound` is `tailIdx`, etc) but decided against because then the method would have grown too big to be manageable and first wanted to run this ideas by you.

If you dim this worth I could extra some of this implementations in submethods and then call them when necessary. 

Fixes #350

Lukasa commented:
@swift-nio-bot test this please

esttorhe commented:
Ugh… just remember that I forgot to run the script for the `Linux` tests 😞 

Incoming

Lukasa commented on Sources/NIO/CircularBuffer.swift:
```diff
@@ -79,6 +85,87 @@ public struct CircularBuffer<E>: CustomStringConvertible, AppendableCollection {
+    /// - Parameter newElements:
+    /// The new elements to add to the collection.
+    ///
+    /// *O(n)* where _n_ is the lenght of the new elements collection if the subrange equals to _n_
```
Nit: `lenght` instead of `length`.

Lukasa commented on Sources/NIO/CircularBuffer.swift:
```diff
@@ -79,6 +85,87 @@ public struct CircularBuffer<E>: CustomStringConvertible, AppendableCollection {
+    ///
+    /// *O(m)* where _m_ is the combined length of the collection and _newElements_
+    public mutating func replaceSubrange(_ subrange: RangeType<Int>, with newElements: ContiguousArray<E>) {
+        guard subrange.lowerBound >= self.startIndex && subrange.upperBound <= self.endIndex else {
```
`endIndex` is actually one step *past* the end of the array, so the upper bound should be closed (e.g. `subrange.upperBound < self.endIndex`.

Lukasa replied:
As a minor nit, may be better to just write this as `precondition`.

Lukasa commented on Sources/NIO/CircularBuffer.swift:
```diff
@@ -79,6 +85,87 @@ public struct CircularBuffer<E>: CustomStringConvertible, AppendableCollection {
+                        newBuffer.append(element)
+                    }
+                    while index != upperBound {
+                        index = self.index(after: index)
```
I *think* you're mixing index types here. I think `headIdx` and `tailIdx` are the internal indices, not the public ones. Instead you probably want `startIndex` and `endIndex`.

Lukasa commented on Sources/NIO/CircularBuffer.swift:
```diff
@@ -79,6 +85,87 @@ public struct CircularBuffer<E>: CustomStringConvertible, AppendableCollection {
+            self.removeSubrange(subrange)
+        } else {
+            var newBuffer: ContiguousArray<E?> = []
+            newBuffer.reserveCapacity(self.buffer.count)
```
This reservation is too small: we can make it 100% accurate by doing the extra bit of math. It's worth doing: if we set the capacity too small we'll have to re-copy all the elements, which is comparatively very expensive.

esttorhe replied:
I'll fix this in a commit within the next hour or so

esttorhe replied:
I'll fix this in a commit within the next hour or so

Lukasa commented on Sources/NIO/CircularBuffer.swift:
```diff
@@ -79,6 +85,87 @@ public struct CircularBuffer<E>: CustomStringConvertible, AppendableCollection {
+    /// *O(n)* where _n_ is the lenght of the new elements collection if the subrange equals to _n_
+    ///
+    /// *O(m)* where _m_ is the combined length of the collection and _newElements_
+    public mutating func replaceSubrange(_ subrange: RangeType<Int>, with newElements: ContiguousArray<E>) {
```
Does any of this function operate on `ContiguousArray` directly? It doesn't seem like it does, in which case it seems like it'd be easier to just make this function generic over `Collection<E>`.

Lukasa commented on Sources/NIO/CircularBuffer.swift:
```diff
@@ -97,17 +184,36 @@ public struct CircularBuffer<E>: CustomStringConvertible, AppendableCollection {
+            while nextIndex != self.tailIdx {
+                self.buffer[bufferIndex] = self.buffer[nextIndex]
+                bufferIndex = nextIndex
+                nextIndex = index(after: bufferIndex)
```
This is calling `index.after` on the wrong entity: you want `self.buffer.index(after)`

Lukasa commented on Sources/NIO/CircularBuffer.swift:
```diff
@@ -205,11 +297,16 @@ public struct CircularBuffer<E>: CustomStringConvertible, AppendableCollection {
     public func index(after: Int) -> Int {
-        let nextIndex = after + 1
-        precondition(nextIndex <= self.endIndex)
+        let nextIndex = (after + 1) >= self.capacity ? 0 : (after + 1)
```
Why did this get changed? It shouldn't have been necessary.

Lukasa commented on Sources/NIO/CircularBuffer.swift:
```diff
@@ -205,11 +297,16 @@ public struct CircularBuffer<E>: CustomStringConvertible, AppendableCollection {
 
+    /// Returns the index before `index`.
+    public func index(before: Int) -> Int {
+        let previousIndex = (before - 1) < 0 ? self.tailIdx : (before - 1)
```
This is mixing indices: it's returning internal indices (`tailIdx`) when it should be returning external ones.

As a separate note, it's also not valid to call `index(before:)` where `before` is `startIndex`, so you can precondition that case away and this degenerates to just a regular subtraction.

Lukasa commented on Sources/NIO/CircularBuffer.swift:
```diff
@@ -52,6 +53,11 @@ public struct CircularBuffer<E>: CustomStringConvertible, AppendableCollection {
+    /// Allocates an empty buffer.
+    public init() {
+        self.buffer = ContiguousArray<E?>()
+    }
```
Do we need the extra `init`?

esttorhe replied:
Unfortunately; it is required by `RangeReplaceableCollection`

Lukasa replied:
Ah, of course it is. That's unfortunate.

weissi replied:
@esttorhe we should only have one `init` function that actually sets the member variables. The empty init should maybe pick a 'sensible' initial ring capacity. What about just doing:

    self.init(initialRingCapacity: 16)

or something?

esttorhe replied:
Makes sense; let me address this in the next commits @weissi

Lukasa commented on Sources/NIO/CircularBuffer.swift:
```diff
@@ -79,6 +85,87 @@ public struct CircularBuffer<E>: CustomStringConvertible, AppendableCollection {
+            var newBuffer: ContiguousArray<E?> = []
+            let capacityDelta = (newElements.count - subrange.count)
+            let newCapacity = Int(UInt32(self.buffer.count + capacityDelta).nextPowerOf2())
+            newBuffer.reserveCapacity(self.buffer.count)
```
You missed using `newCapacity`.

esttorhe replied:
shouldn't be coding at 00:13; let me address the feedback in the next commit

Lukasa commented:
@swift-nio-bot test this please

Lukasa commented on Sources/NIO/CircularBuffer.swift:
```diff
@@ -79,6 +85,87 @@ public struct CircularBuffer<E>: CustomStringConvertible, AppendableCollection {
+            newBuffer.append(contentsOf: repeatElement(nil, count: self.count))
+            self.headIdx = 0
+            self.tailIdx = 0
+            self.buffer = newBuffer
```
You can do this whole block with `self = .init(initialRingCapacity: self.buffer.count)`

Lukasa commented on Sources/NIO/CircularBuffer.swift:
```diff
@@ -151,6 +243,18 @@ public struct CircularBuffer<E>: CustomStringConvertible, AppendableCollection {
 
+    /// Returns the internal buffer next index after `index`.
+    private func bufferIndex(after: Int) -> Int {
+        let nextIndex = (after + 1) >= self.capacity ? 0 : (after + 1)
```
Rather than use the ternary, maybe just use the masking math that we do elsewhere?

Lukasa commented on Sources/NIO/CircularBuffer.swift:
```diff
@@ -151,6 +243,18 @@ public struct CircularBuffer<E>: CustomStringConvertible, AppendableCollection {
+
+    /// Returns the internal buffer index before `index`.
+    private func bufferIndex(before: Int) -> Int {
+        let previousIndex = (before - 1) < 0 ? self.tailIdx : (before - 1)
```
Same note about the ternary here.

Lukasa commented on Sources/NIO/CircularBuffer.swift:
```diff
@@ -206,10 +310,17 @@ public struct CircularBuffer<E>: CustomStringConvertible, AppendableCollection {
     public func index(after: Int) -> Int {
         let nextIndex = after + 1
-        precondition(nextIndex <= self.endIndex)
+    	precondition(nextIndex <= self.endIndex)
```
No need for this change.

Lukasa commented on Sources/NIO/CircularBuffer.swift:
```diff
@@ -79,6 +85,87 @@ public struct CircularBuffer<E>: CustomStringConvertible, AppendableCollection {
+                    for element in newElements {
+                        newBuffer.append(element)
+                    }
+                    while index != upperBound {
```
As a simple correctness note it may be better to do `while index < upperBound`, to tolerate the possibility of zero-sized ranges.

Lukasa commented on Sources/NIO/CircularBuffer.swift:
```diff
@@ -79,6 +85,87 @@ public struct CircularBuffer<E>: CustomStringConvertible, AppendableCollection {
+    /// *O(n)* where _n_ is the length of the new elements collection if the subrange equals to _n_
+    ///
+    /// *O(m)* where _m_ is the combined length of the collection and _newElements_
+    public mutating func replaceSubrange<C>(_ subrange: RangeType<Int>, with newElements: C) where C : Collection, E == C.Element {
```
The build failed because you shouldn't use `RangeType` here, just `Range`.

Lukasa commented on Tests/NIOTests/CircularBufferTests.swift:
```diff
@@ -125,12 +136,30 @@ class CircularBufferTests: XCTestCase {
+            ring.prepend(idx)
+        }
+        XCTAssertEqual(50, ring.count)
+        let newElements = ContiguousArray<Int>(arrayLiteral: 99)
```
You shouldn't use the `arrayLiteral` initialiser directly, instead just do `let newElements: ContiguousArray<Int> = [99]`

Lukasa replied:
And in fact given that you're using `replaceSubrange` in this test you can just use a regular array.

Lukasa commented on Tests/NIOTests/CircularBufferTests.swift:
```diff
@@ -125,12 +136,29 @@ class CircularBufferTests: XCTestCase {
+        XCTAssertEqual(ring[21], 24)
+        ring.replaceSubrange(ring.startIndex..<ring.endIndex, with: [])
+        XCTAssertTrue(ring.isEmpty)
+    }
```
Can this test be extended to also test `replaceSubrange` with a subrange that is *larger* than the target range, and one that is equal to the subrange? Would be good to cover all the cases.

Lukasa commented on Sources/NIO/CircularBuffer.swift:
```diff
@@ -79,6 +85,81 @@ public struct CircularBuffer<E>: CustomStringConvertible, AppendableCollection {
+        }
+    }
+
+    public mutating func removeSubrange(_ bounds: Range<Int>) {
```
This should probably precondition that the `bounds` are in fact valid indices for the collection.

Lukasa replied:
Can you also add a doc comment here?

Lukasa commented on Sources/NIO/CircularBuffer.swift:
```diff
@@ -79,6 +85,81 @@ public struct CircularBuffer<E>: CustomStringConvertible, AppendableCollection {
+        }
+    }
+
+    public mutating func removeLast(_ n: Int) {
```
Can you add a doc comment here too please?

Lukasa replied:
Also this should validate that the value of `n` is reasonable.

esttorhe commented:
Feedback addressed; ready for another test run and some 👀 🙇

Lukasa commented on Sources/NIO/CircularBuffer.swift:
```diff
@@ -79,6 +85,81 @@ public struct CircularBuffer<E>: CustomStringConvertible, AppendableCollection {
+        self.tailIdx = (self.tailIdx - n) & self.mask
+        for index in end - n ..< end {
+            self.buffer[index] = nil
+        }
```
Can you add a test for trying to use `removeLast` while the buffer is non-contiguous? I think this code doesn't work because it assumes that the indices are contiguous, which they are not.

Lukasa commented on Sources/NIO/CircularBuffer.swift:
```diff
@@ -97,17 +178,36 @@ public struct CircularBuffer<E>: CustomStringConvertible, AppendableCollection {
+    /// *O(1)* if the position is `headIdx` or `tailIdx`.
+    /// otherwise
+    /// *O(n)* where *n* is the number of elements between `position` and `tailIdx`.
+    public mutating func remove(at position: Int) -> E {
```
This should probably validate `position` is valid.

Lukasa commented on Sources/NIO/CircularBuffer.swift:
```diff
@@ -97,17 +178,36 @@ public struct CircularBuffer<E>: CustomStringConvertible, AppendableCollection {
+        case self.headIdx:
+            self.headIdx = (self.headIdx + 1) & self.mask
+            self.buffer[bufferIndex] = nil
+        case self.tailIdx - 1:
```
This looks wrong to me: can you test this function including this case?

esttorhe replied:
Test can be found here https://github.com/esttorhe/swift-nio/blob/9133c12f9a7122741c0d8ae750e6a93d5bd7be83/Tests/NIOTests/CircularBufferTests.swift#L65

Lukasa commented:
@swift-nio-bot test this please

esttorhe commented:
Addressed the feedback.
Simplified the `replaceSubrange` logic and added tests for all potential paths to guarantee that no breaking changes in the future

Lukasa commented on Sources/NIO/CircularBuffer.swift:
```diff
@@ -79,6 +85,77 @@ public struct CircularBuffer<E>: CustomStringConvertible, AppendableCollection {
+
+            newBuffer.append(contentsOf: self[0..<subrange.lowerBound].map { $0 })
+            newBuffer.append(contentsOf: newElements.map { $0 })
+            newBuffer.append(contentsOf: self[subrange.upperBound..<self.endIndex].map { $0 })
```
`append(contentsOf:)` takes `Sequence`, so there should be no need for the call to `map`.

esttorhe replied:
That was my first idea but the compiler doesn't let me and complains that the `contentsOf:` label is not needed 😕; which obviously is wrong because then I tries to append the sequence as if it was an array of sequences

esttorhe replied:
<img width="1245" alt="circularbuffer swift 2018-06-06 19-12-08" src="https://user-images.githubusercontent.com/2642850/41054051-961cbf0e-69bd-11e8-9713-3639add03b76.png">

esttorhe replied:
It makes sense though… because I'm slicing `CircularBuffer` and then try to append to `ContiguosArray` so I end up trying to add `Slice<CircularBuffer<E>>` to `ContiguosArray<E>`

Lukasa replied:
Well that's annoying. I just filed [SR-7921](https://bugs.swift.org/browse/SR-7921) to resolve the fact that this fixes the issue.

In the meantime, please update this with a comment indicating that this behaviour is silly.

Lukasa commented on Sources/NIO/CircularBuffer.swift:
```diff
@@ -97,17 +174,37 @@ public struct CircularBuffer<E>: CustomStringConvertible, AppendableCollection {
+
+        switch bufferIndex {
+        case self.headIdx:
+            self.headIdx = (self.headIdx + 1) & self.mask
```
Now you have `bufferIndex(after:)` you can rewrite this to use `self.headIdx = self.bufferIndex(after: self.headIdx)`. It reads a bit more nicely. :)

esttorhe replied:
Nice 👍

Lukasa commented:
@swift-nio-bot test this please

esttorhe commented:
@Lukasa comment added

Lukasa commented:
@swift-nio-bot test this please

esttorhe commented:
Should I click the `Update branch` button?

BasThomas commented:
Probably fine to leave it for the person that merges the pull request; don’t worry. 

<sub>Sent with <a href="http://githawk.com">GitHawk</a></sub>

esttorhe commented:
Makes sense. Thanks

On Sat 9. Jun 2018 at 13:46, Bas Broek <notifications@github.com> wrote:

> Probably fine to leave it for the person that merges the pull request;
> don’t worry.
>
> Sent with GitHawk <http://githawk.com>
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/apple/swift-nio/pull/466#issuecomment-395962878>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AChTok7eouN1iwdjW9jbwGGP3x7gTW5eks5t67WEgaJpZM4UY14n>
> .
>
-- 
—
Esteban Torres
iOS Developer

Phone: (+49)0162-4190-543
Skype: esteban.a.torres.hernandez
Twitter: @esttorhe

Lukasa commented:
@esttorhe just a gentle reminder that there's some code review outstanding here. :smile:

esttorhe commented:
@Lukasa yeah I saw it, all the feedback seems fair. 
I'm on holidays and that's why I haven't been able to address it. I'll do it later today or tomorrow night. 

Thanks

esttorhe commented:
@Lukasa, @normanmaurer just made the requested changes 🙂

Lukasa commented on Sources/NIO/CircularBuffer.swift:
```diff
@@ -79,6 +85,80 @@ public struct CircularBuffer<E>: CustomStringConvertible, AppendableCollection {
+        precondition(subrange.lowerBound >= self.startIndex && subrange.upperBound <= self.endIndex, "Subrange out of bounds")
+
+        if subrange.count == newElements.count {
+            for (index, element) in zip(stride(from: subrange.lowerBound, to: subrange.upperBound, by: 1), newElements) {
```
The `stride` here is not necessary, the `Range` is a sequence of the correct type already. This can just be `zip(subrange, newElements)`.

esttorhe replied:
Can't do it just like that because the compiler complains with:
```argument type 'Range<Int>' does not conform to expected type 'Sequence'```

I changed it to 
```swift
zip(subrange.lowerBound..<subrange.upperBound, newElements)
```

I think this reads better than using `stride`. Also added a comment pointing out why is not using just `zip(subrange, newElements)`

Lukasa commented on Sources/NIO/CircularBuffer.swift:
```diff
@@ -79,6 +85,80 @@ public struct CircularBuffer<E>: CustomStringConvertible, AppendableCollection {
+            let repetitionCount = self.buffer.count - newBuffer.count
+            if repetitionCount > 0 {
+                newBuffer.append(contentsOf: repeatElement(nil, count: repetitionCount))
+            }
```
This math is wrong. The following test fails, but should succeed:

```swift
    func testBufferManaged() {
        var ring = CircularBuffer<Int>(initialRingCapacity: 2)
        ring.append(1)
        XCTAssertEqual(ring.capacity, 2)

        // Now we want to replace the last subrange with two elements. This should
        // force an increase in size.
        ring.replaceSubrange(0..<1, with: [3, 4])
        XCTAssertEqual(ring.capacity, 4)
    }
```

The core issue here is that you are filling up to the *old* buffer size, but you need to fill up to the *new* one. `repetitionCount` should be calculated as `newCapacity - newBuffer.count`. That will also allow us to avoid a branch, as the number can never be negative: just append `repeatElement` unconditionally.

esttorhe replied:
Ugh!! shame cube.

Good catch Cory. I'll fix it during the day
Thanks

weissi commented on Sources/NIO/CircularBuffer.swift:
```diff
@@ -79,6 +85,80 @@ public struct CircularBuffer<E>: CustomStringConvertible, AppendableCollection {
+            #if swift(>=4.2)
+            let capacityDelta = (newElements.count - subrange.count)
+            #else
+            let capacityDelta = (Int(newElements.count) - subrange.count)
```
the extra `Int(...)` doesn't hurt, can we please remove the `#if` dance?

weissi commented on Sources/NIO/CircularBuffer.swift:
```diff
@@ -23,7 +23,8 @@ public protocol AppendableCollection: Collection {
 /// will automatically expand if more elements than `initialRingCapacity` are stored, it's advantageous to prevent
 /// expansions from happening frequently. Expansions will always force an allocation and a copy to happen.
-public struct CircularBuffer<E>: CustomStringConvertible, AppendableCollection {
+public struct CircularBuffer<E>: CustomStringConvertible, AppendableCollection, BidirectionalCollection, RandomAccessCollection, RangeReplaceableCollection
```
could we move all the new conformances that are implementable using the public/internal API to extensions? I think the actual `struct CircularBuffer` should be as concise as possible

weissi commented on Sources/NIO/CircularBuffer.swift:
```diff
@@ -79,6 +85,80 @@ public struct CircularBuffer<E>: CustomStringConvertible, AppendableCollection {
+            // This mapping is required due to an inconsistent ability to append sequences of non-optional
+            // to optional sequences.
+            // https://bugs.swift.org/browse/SR-7921
+            newBuffer.append(contentsOf: self[0..<subrange.lowerBound].map { $0 })
```
these should all be `.lazy` maps, no?

weissi commented on Tests/NIOTests/CircularBufferTests.swift:
```diff
@@ -125,12 +145,47 @@ class CircularBufferTests: XCTestCase {
         XCTAssertEqual(expectedValues, actualValues)
     }
 
+    func testReplacingSubrangeOfElements() {
```
- we should split this up into appropriately named test cases
- we need more test cases here that deliberately test some edge conditions like ones that wrap over
- have you checked the code coverage?

weissi commented on Tests/NIOTests/CircularBufferTests.swift:
```diff
@@ -266,6 +321,28 @@ class CircularBufferTests: XCTestCase {
         XCTAssertNil(ring.last)
         XCTAssertEqual(0, ring.count)
         XCTAssertTrue(ring.isEmpty)
+        ring.prepend(9)
```
these should be new test cases rather than modifying existing ones

weissi commented on Tests/NIOTests/CircularBufferTests.swift:
```diff
@@ -51,6 +51,26 @@ class CircularBufferTests: XCTestCase {
     }
 
+    func testRemoveAt() {
+        var ring = CircularBuffer<Int>(initialRingCapacity: 4)
```
separate test cases.

weissi commented:
ping @esttorhe, almost there here and it'd be great to ship with 1.9.0 which we will be shipping soonish I guess (we don't have a precise date yet).

esttorhe commented:
Hi Johannes,

Definitely need to wrap the last batch of changes.
Unfortunately I was shipping the PR while in-between quarters planning and
then I got lots of work on me.

I'll try to find some time this week to address the last round of feedback.

On Fri 6. Jul 2018 at 13:02, Johannes Weiss <notifications@github.com>
wrote:

> ping @esttorhe <https://github.com/esttorhe>, almost there here and it'd
> be great to ship with 1.9.0 which we will be shipping soonish I guess (we
> don't have a precise date yet).
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/apple/swift-nio/pull/466#issuecomment-403002402>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AChTog9ja4IZmAWqIX3oddzys4BmxCWlks5uD0OrgaJpZM4UY14n>
> .
>
-- 
—
Esteban Torres
iOS Developer

Phone: (+49)0162-4190-543
Skype: esteban.a.torres.hernandez
Twitter: @esttorhe

weissi commented on Sources/NIO/CircularBuffer.swift:
```diff
@@ -23,7 +23,8 @@ public protocol AppendableCollection: Collection {
 /// will automatically expand if more elements than `initialRingCapacity` are stored, it's advantageous to prevent
 /// expansions from happening frequently. Expansions will always force an allocation and a copy to happen.
-public struct CircularBuffer<E>: CustomStringConvertible, AppendableCollection {
+public struct CircularBuffer<E>: CustomStringConvertible, AppendableCollection
```
mind taking this change out, we generally have the curly braces on the same line

weissi commented on Sources/NIO/CircularBuffer.swift:
```diff
@@ -210,6 +175,12 @@ public struct CircularBuffer<E>: CustomStringConvertible, AppendableCollection {
+    /// Returns the index before `index`.
+    public func index(before: Int) -> Int {
+        precondition(before > 0)
+        return (before - 1)
```
don't need the parens

Lukasa commented on Sources/NIO/CircularBuffer.swift:
```diff
@@ -234,3 +204,131 @@ public struct CircularBuffer<E>: CustomStringConvertible, AppendableCollection {
+        case self.headIdx:
+            self.headIdx = self.bufferIndex(after: self.headIdx)
+            self.buffer[bufferIndex] = nil
+        case self.tailIdx - 1:
```
Does this handle the case where `tailIdx` is 0?

normanmaurer commented:
@Lukasa yep will do later today.

normanmaurer commented on Sources/NIO/CircularBuffer.swift:
```diff
@@ -234,3 +182,131 @@ public struct CircularBuffer<E>: CustomStringConvertible, AppendableCollection {
+            _ = remove(at: bounds.lowerBound)
+        } else if bounds.count == self.count {
+            self = .init(initialRingCapacity: self.buffer.count)
+        } else {
```
nit: should we use a switch here ?

esttorhe replied:
I think a `switch` would make this a lot more readable.

I'll change it during the day

Lukasa commented:
Mind bringing this up to date with the current master branch?

esttorhe commented:
@lukasa, I usually rebase with master, but that would require force pushing and rewriting history. 
Or do you want me to just merge master into my branch?

Lukasa commented:
Feel free to rebase on master and then force push.

esttorhe commented:
@Lukasa done. For whatever reason git kept telling me that it was outdated; so I ended up just clicking the `Update branch` button (but I did cleaned up the history a bit)

weissi commented:
@esttorhe just in case you need it next time:

```
git fetch upstream
git rebase upstream/master
git push -f esttorhe issue_350
```

with the following assumptions:
- your git remote that points to github.com/apple/swift-nio is named `upstream`
- your git remote that points to github.com/esstorhe/swift-nio is named `esttorhe`
- your branch is named `issue_350`

you can find the names of the remotes out using `git remote -v`. For me for example it's

```
$ git remote -v
jw	git@github.com:weissi/swift-nio.git (fetch)
jw	git@github.com:weissi/swift-nio.git (push)
upstream	git@github.com:apple/swift-nio.git (fetch)
upstream	READ-ONLY (push)
```

so for me it'd be

```
git fetch upstream
git rebase upstream/master
git push -f jw issue_350
```

esttorhe commented:
Yeah; I have a handy `alias` that does all of this already; the problem is that I'm used to having `origin` as the; well; origin 😋 

But in this case `origin` points to my `fork` and `source` is this repository… so of course every time that I tried rebasing I was rebasing with my outdated `master` from `origin` 😞

Lukasa commented:
@swift-nio-bot test this please

BasThomas commented:
Congrats on shipping this @esttorhe! 😀

esttorhe commented:
Thanks!

Happy is merged.
Now to grab another task 😊

On Thu 12. Jul 2018 at 09:10, Bas Broek <notifications@github.com> wrote:

> Congrats on shipping this @esttorhe <https://github.com/esttorhe>! 😀
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/apple/swift-nio/pull/466#issuecomment-404412183>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AChTouQqmtZRnBrVSptbt1HKHBxGHHf9ks5uFvZSgaJpZM4UY14n>
> .
>
-- 
—
Esteban Torres
iOS Developer

Phone: (+49)0162-4190-543
Skype: esteban.a.torres.hernandez
Twitter: @esttorhe
